### PR TITLE
BUG: Make `mask_invalid` consistent with `mask_where` if `copy` is set to `False`

### LINKED ISSUE #19332: BUG: ma.masked_invalid(a, copy=False) does not modify a
According to [`masked_invalid` docs](https://numpy.org/doc/stable/reference/generated/numpy.ma.masked_invalid.html), it is a shortcut for [`masked_where`](https://numpy.org/doc/stable/reference/generated/numpy.ma.masked_where.html), which says that `copy=False` will modify `a` in place.

But `masked_where` didn't quite do this, and was recently changed in 1.21 to match what the docs actually say in #18967. While it seems to have propagated to e.g., `masked_less`, it has not affected `masked_invalid`, which continues to _not_ modify `a`.

Note though, that unlike `masked_where`, a partially masked input array does not 'fix' `copy=False` to match the docs, as in #18946. So changing this behaviour might be more problematic and it might be better to make this explicit in the docs instead.

### Reproducing code example:

`masked_where` modifies input:
```python
In [1]: import numpy as np

In [25]: a = np.ma.array([np.inf, 1, 2, 3, 4])

In [26]: a
Out[26]: 
masked_array(data=[inf,  1.,  2.,  3.,  4.],
             mask=False,
       fill_value=1e+20)

In [28]: np.ma.masked_where(a == 3, a, copy=False)
Out[28]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=1e+20)

In [29]: a
Out[29]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=1e+20)
```

Other wrappers like `masked_equal` modify input:
```python
In [30]: a = np.ma.array([np.inf, 1, 2, 3, 4])

In [31]: np.ma.masked_equal(a, 3, copy=False)
Out[31]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=3.0)

In [32]: a
Out[32]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=1e+20)
```

But `masked_invalid` does not:
```python
In [33]: a = np.ma.array([np.inf, 1, 2, 3, 4])

In [34]: np.ma.masked_invalid(a, copy=False)
Out[34]: 
masked_array(data=[--, 1.0, 2.0, 3.0, 4.0],
             mask=[ True, False, False, False, False],
       fill_value=1e+20)

In [35]: a
Out[35]: 
masked_array(data=[inf,  1.,  2.,  3.,  4.],
             mask=False,
       fill_value=1e+20)
```

And unlike `masked_where` before, it does _not_ magically work again for a partially masked array:
```
In [36]: a = np.ma.array([np.inf, 1, 2, 3, 4], mask=[False, False, False, True, False])

In [37]: a
Out[37]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=1e+20)

In [38]: np.ma.masked_invalid(a, copy=False)
Out[38]: 
masked_array(data=[--, 1.0, 2.0, --, 4.0],
             mask=[ True, False, False,  True, False],
       fill_value=1e+20)

In [39]: a
Out[39]: 
masked_array(data=[inf, 1.0, 2.0, --, 4.0],
             mask=[False, False, False,  True, False],
       fill_value=1e+20)
```

### NumPy/Python version information:
```
1.21.0 3.7.6 | packaged by conda-forge | (default, Mar  5 2020, 15:27:18) 
[GCC 7.3.0]
```

### PR DESCRIPTION
**Pull request by cmarmo**
This pull requests makes `mask_invalid` consistent with `mask_where` when `copy` is set to `False`.
Fixes #19332.

I'd rather make the two functions consistent than change the documentation.
Also from the documentation

> Only applies to arrays with a dtype where NaNs or infs make sense

I added a test checking that an error is thrown when `isinfinite` is not applicable.

Thanks for considering it.

### CODE REVIEW

File: numpy/ma/core.py [hunk 1]
```diff
@@ -2356,20 +2356,11 @@ def masked_invalid(a, copy=True):
...
-        condition = ~(np.isfinite(a))
-        cls = MaskedArray
-    result = a.view(cls)
-    result._mask = condition
-    return result
 
+    try:
+        return masked_where(~(np.isfinite(getdata(a))), a, copy=copy)
+    except TypeError:
+        raise
```
Comment (bsipocz) [inline numpy/ma/core.py:2363]: Was there a particular reason for this try/except, or is it a debug remnant?
```suggestion
    return masked_where(~(np.isfinite(getdata(a))), a, copy=copy)
```

File: numpy/ma/tests/test_core.py [hunk 2]
```diff
@@ -4496,6 +4496,14 @@ def test_where_structured_masked(self):
...
         assert_equal(ma.mask, expected.mask)
 
+    def test_masked_invalid_error(self):
+        a = np.arange(5, dtype=object)
+        a[3] = np.PINF
+        a[2] = np.NaN
+        with pytest.raises(TypeError,
+                           match="not supported for the input types"):
+            np.ma.masked_invalid(a)
+
```
Comment (seberg) [inline numpy/ma/tests/test_core.py:4506]: We were looking at this in the triage meeting as well.  I guess the test is great (I honestly still need to parse it fully).
But we are missing an additional test for the successful path that was fixed I think:  Checking that the array was indeed modified in-place with `copy=False`?

Comment (seberg) [reply numpy/ma/tests/test_core.py:4506]: Yeah, thanks.  It is true that `masked_where` is tested, but it is also just nice to see the fix in action in the PR and we mainly have integration tests anyway.

A test too much cannot hurt :).

I find it odd that we consider `inf` invalid by default, but that is not a change.

Thanks @cmarmo, lets get this in!

**Comment by seberg** [comment]
I suppose we should do this just to align things anyway...

Although, now I actually suspect that this may have been intentional (Chesterton's fence greets):  The function always replaces the mask, but does not always copy the data, which is a pattern to the madness that does make sense, the docs are probably just fuzzy on that intention (if it was the intention).
Sorry, long shot, but @ahaldane do you happen to have a gut feeling on this?

**Comment by InessaPawson** [comment]
@mhvk Do you have any thoughts on this?

**Comment by mhvk** [comment]
Not really. Overall, to me this PR makes sense: do as the doc states and just call `masked_where`. I've never understood why if data is kept, masks are not, though clearly it was designed that way.

**Comment by seberg** [comment]
Thanks @cmarmo and Marten, lets give this a shot then!

**Comment by charris** [comment]
Should probably add a release note for this.

**Comment by bsipocz** [comment]
Hmm, if the label is used consistently, a script could help to look for missing release notes before release time?

**Comment by cmarmo** [comment]
>  @cmarmo are you interested in having a look at that, the instructions are in `numpy/doc/release/upcoming_changes/README.txt` (you basically add a file in that folder with a `22046.change.rst` as name).

I can take care of that if it can wait until the week-end. :)

**Pull request merged**